### ISSUE TITLE
FR: Use terraform syntax highlighting for terraform template .tftpl files

### ISSUE THREAD

**Issue by AdamSLevy** [hashivim/vim-terraform#176]
Terraform template files, typically ending in `.tftpl` could benefit from the terraform syntax highlighting.

**Comment by dimbleby** [comment]
I don't think this is true, a template file can contain arbitrary text and not usually HCL.

Did I misunderstand?

**Comment by radeksimko** [comment]
Template files which are passed into [`templatefile()`](https://www.terraform.io/language/functions/templatefile) have to contain valid HCL/Terraform _expression_, that is equivalent to 

```hcl
attribute = EXPRESSION_HERE
```
The only difference from the standard config ^ is that the file contents is always implied to be a string and it doesn't have to be enclosed in double quotes. Therefore, this

```hcl
attribute = "arbitrary text"
```
is equivalent to this
```hcl
attribute = templatefile("template.tftpl", {})
```
**`template.tftpl`**
```hcl
arbitrary text
```
So you are technically correct in that arbitrary text is valid HCL expression and therefore `*.tftpl` _may_ contain arbitrary text with no interpolation syntax, but any interpolation logic which it _does_ contain _has to be valid_. For example, if your file contains `some ${unknown-variable} text` or even just `${` then this would be considered invalid and rejected unless properly escaped. Additionally, templates also accept [directives (`%{}`)](https://www.terraform.io/language/expressions/strings#directives) which would also have to be correctly escaped, if you're passing them into `templatefile()`.

Either way in case folks don't need interpolation, they would use [`file()`](https://www.terraform.io/language/functions/file) instead, so while you _can_, you _usually_ would _not_ keep arbitrary text in `*.tftpl`.

TL;DR I think that highlighting `*.tftpl` is a relevant ask.

How to best highlight it though is another question, because it is technically _not_ equivalent to `*.tf` nor `*.tfvars`. Ideally it should be treated as the "inside the double quotes" after equal sign. I am not familiar with vim highlighting engine myself to suggest what exactly does that mean there.

**Comment by dimbleby** [comment]
I think we agree that it's not useful to highlight template files in the same way as terraform files.  

I'm sceptical about the value of adding specialised highlighting just to deal with the templating, all it is going to buy is a small amount of string interpolation in an otherwise arbitrary file type.  Such things are possible, typically you end up with compound filetypes like `html.mustache` or whatever, I am just sceptical about the value in this case.

Anyway: I have no plans to add highlighting dedicated to template files and so will leave this closed.  But if anyone shows up with an implementation and a convincing argument that it's useful, I'll be glad to accept the MR.

**Comment by radeksimko** [comment]
> Such things are possible, typically you end up with compound filetypes like `html.mustache` or whatever, I am just sceptical about the value in this case.

I understand your sentiment. The big question [for me] would be more generally how vim highlighting deals with embedded (typically templating) languages. That then helps answer the effort vs value question. Having faced the same problem in VS Code, I know it's not trivial there and I don't expect it to be trivial elsewhere.

btw. codeblocks within Markdown files face a very similar problem - so if I was looking for a solution, that's where I would look.

**Comment by dimbleby** [comment]
that's kind of the opposite problem, or anyway some sort of dual version of it

- There you definitely know that you have a markdown file, and something arbitrary embedded in it
- here it's more like you have mostly some arbitrary file, and a small amount of known templating syntax embedded in it

so I guess that comparison with how plugins for eg jinja2 and handlebars and similar work would be more useful for anyone who does want to work on this.

But it doesn't sound like that's either of us!
